File: BAEKJOON/카테고리/DFS_BFS/1260_DFS와_BFS/1260.py
import sys

sys.stdin = open('input.txt', 'r')


def DFS(start, graph):
    stack = [start]

    visited = []

    while stack:
        node = stack.pop(-1)

        if node not in visited:

            visited.append(node)

            # 작은 순서대로 가기 위해서 정렬 -> 뒤집기가 필요함 (뒤에서 빼니까)
            childes = sorted(list(graph[node]))

            childes.reverse()

            for child in childes:

                if child not in visited:
                    stack.append(child)

    return visited


def BFS(start, graph):
    q = [start]

    visited = []

    while q:
        node = q.pop(0)

        if node not in visited:

            visited.append(node)

            # 작은 순서대로 가기 위해서 정렬이 필요함 (앞에서 빼니까)
            childes = sorted(list(graph[node]))

            for child in childes:

                if child not in visited:
                    q.append(child)
    return visited


N, M, V = map(int, input().split())

graph = {}

for i in range(1, N + 1):
    graph[i] = set()

for _ in range(M):
    node1, node2 = map(int, input().split())

    graph[node1].add(node2)
    graph[node2].add(node1)

print(" ".join(map(str, DFS(V, graph))))
print(" ".join(map(str, BFS(V, graph))))

File: javascript/BOJ/1260_DFS와BFS/1260.js
const fs = require('fs');
const fileStream = fs.createReadStream('input.txt');
const readline = require('readline');

const rl = readline.createInterface({
  input: fileStream,
  output: process.stdout,
})

let N = null;
let M = null;
let V = null;
let i = 0;
const graph = {};

function DFS(x) {
  const stack = [x];
  const visited = [];
  while (stack.length > 0) {
    const node = stack.pop();

    if (!visited.includes(node)) {
      visited.push(node)
      for (let c of [...graph[node]].sort((a, b) => a > b ? -1 : 1)) {
        if (!visited.includes(c)) {
          stack.push(c);
        }
      }
    }
  }
  return visited
}

function BFS(x) {
  const queue = [x];
  const visited = [];

  while (queue.length > 0) {
    const node = queue.shift();
    if (!visited.includes(node)) {
      visited.push(node)
      for (let c of [...graph[node]].sort()) {
        if (!visited.includes(c)) {
          queue.push(c);
        }
      }
    }
  }
  return visited
}

rl.on("line", (line) => {
  if (i === 0) {
    [N, M, V] = line.split(" ").map(Number);
    for (let i = 1; i <= N; i++) {
      graph[i] = [];
    }
  } else {
    const [v1, v2] = line.split(" ").map(Number);
    graph[v1].push(v2)
    graph[v2].push(v1)
  }
  i++
})
  .on("close", () => {
    console.log(DFS(V).join(" "))
    console.log(BFS(V).join(" "))
    process.exit()
  })